### DISCUSSION TITLE
What the difference between stream and stream_log?

### DISCUSSION THREAD

**Discussion by rever1and** [langchain-ai/langserve#431] [Q&A] [answered]
When to use which one?

**Comment by eyurtsev** [comment, marked as answer]
It depends on which level of granularity / customization you need.

stream gives a default implementation and stream only the final output. 

if you need to stream intermediate steps, you can use stream_log, though we now have a newer API called stream events which might be easier.

See this documentation: https://python.langchain.com/docs/expression_language/streaming

**Reply by rever1and** [reply]
Thanks, during these days I have migrated to stream_events, it nearly have all things one need, without needing to parse agent result and can streaming tool intermediate result easy.

**Reply by eyurtsev** [reply]
Is there anything that's missing that you need?

**Reply by rever1and** [reply]
no, I think I've got everything I need. Thanks_
